[gcc] Define '--target' matching operating system

stephenmsachs opened the pull request:
The GNU target triplet is not always guessed correctly by the configure script. Use the pre-existing `gcc` to define the correct build target.

This fixes https://github.com/spack/spack/issues/27668 as the Intel compiler needs the correct Debian tuplet `x86_64-linux-gnu` as `target` in its underlying `gcc` to find files under `/usr/include/x86_64-linux-gnu`.

I looked into using the build scripts of the different `gcc` packages in Linux distributions, but there are just too many differences, e.g. Debian based distributions use their own triplets (https://wiki.debian.org/Multiarch/Tuples). As `gcc` is a dependency for Spack installation, we can safely assume that `existing_gcc` exists and can be used to report its build target.

Commit:
[gcc] Define '--target' matching operating system

The GNU target triplet is not always guessed correctly by the configure script.
Use the pre-existing `gcc` to define the correct build target.

This fixes https://github.com/spack/spack/issues/27668 as the Intel compiler
needs the correct Debian tuplet `x86_64-linux-gnu` as `target` in its underlying
`gcc` to find files under `/usr/include/x86_64-linux-gnu`.

I looked into using the build scripts of the different `gcc` packages in Linux
distributions, but there are just too many differences, e.g. Debian based
distributions use their own triplets (https://wiki.debian.org/Multiarch/Tuples).
As `gcc` is a dependency for Spack installation, we can safely assume that
`existing_gcc` exists and can be used to report its build target.

wyphan commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -755,6 +755,13 @@ def configure_args(self):
         ]
 
+        # Re-use the GNU target triplet defined in the existing gcc
+        exisitng_gcc = Executable("gcc")
```
```suggestion
        existing_gcc = Executable("gcc")
```

wyphan commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -755,6 +755,13 @@ def configure_args(self):
 
+        # Re-use the GNU target triplet defined in the existing gcc
+        exisitng_gcc = Executable("gcc")
+        triplet = exisitng_gcc("-###", error=str).partition("Target: ")[2].split("\n")[0]
```
```suggestion
        triplet = existing_gcc("-###", error=str).partition("Target: ")[2].split("\n")[0]
```

stephenmsachs committed:
fix typo

Co-authored-by: Wileam Y. Phan <50928756+wyphan@users.noreply.github.com>

stephenmsachs committed:
fix typo

Co-authored-by: Wileam Y. Phan <50928756+wyphan@users.noreply.github.com>

michaelkuhn commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -755,6 +755,13 @@ def configure_args(self):
             "--disable-nls",
         ]
 
+        # Re-use the GNU target triplet defined in the existing gcc
+        existing_gcc = Executable("gcc")
+        triplet = existing_gcc("-###", error=str).partition("Target: ")[2].split("\n")[0]
+        if triplet:
+            options.append("--build={}".format(triplet))
+            options.append("--target={}".format(triplet))
```
Should we wrap this in `if spec.satisfies("%gcc")` and use `self.compiler.cc` to be extra sure (since `gcc` might not actually be the compiler we build GCC with)?

Moreover, shouldn't we also set `--host` to the same value (see https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html)?

alalazo replied:
> Use the pre-existing gcc to define the correct build target.

Not sure if this is the way to go. Wouldn't this block us from having cross-compilers (which is something we want to do in the mid-term)? 

> This fixes https://github.com/spack/spack/issues/27668 as the Intel compiler needs the correct Debian tuplet x86_64-linux-gnu as target in its underlying gcc to find files under /usr/include/x86_64-linux-gnu.

This seems more an issue with Intel than with GCC. Can't we fix it in the Intel compiler class?

alalazo reviewed (changes requested):
> As gcc is a dependency for Spack installation, we can safely assume that existing_gcc exists and can be used to report its build target.

I think this is a false assumption in general. We never stated that gcc is a requirement for Spack

stephenmsachs commented:
I am testing Michaels suggestion to use `if spec.satisfies("%gcc")` right now. so this should not be a problem. Buit the packages listed in "System Prerequisites" in the [Getting Started Guide](https://spack.readthedocs.io/en/latest/getting_started.html) will install `gcc` in all cases.

alalazo commented:
@stephenmsachs Can you expand a bit on this:

> This fixes https://github.com/spack/spack/issues/27668 as the Intel compiler needs the correct Debian tuplet x86_64-linux-gnu as target in its underlying gcc to find

in the issue #27668 ? From reading it it is not clear why this would be the solution, so I think adding the missing bits of information in the issue can help getting the context quickly.

stephenmsachs commented:
> @stephenmsachs Can you expand a bit on this:
> 
> > This fixes https://github.com/spack/spack/issues/27668 as the Intel compiler needs the correct Debian tuplet x86_64-linux-gnu as target in its underlying gcc to find
> 
> in the issue #27668 ? From reading it, it is not clear why this would be the solution, so I think adding the missing bits of information in the issue can help getting the context quickly.

OK, I'll try to my best knowledge. This is the summary of the thoughts on https://github.com/spack/spack/pull/27669.

1. On Ubuntu some of the standard standard include files (e.g. `bits/types.h`) are under `/usr/include/x86_64-linux-gnu`. On RHEL those would be found directly under `/usr/include`. 

2. Also, on Ubuntu `spack install gcc` will install `gcc` with Target: `x86_64-pc-linux-gnu`. The `pc` is the important part.

3.  If you install the Intel compiler (which needs `gcc` as dependency) on Ubuntu it will not be able to find the include files under `/usr/include/x86_64-linux-gnu`.

I have found out that if the `gcc` is installed with Target `x86_64-linux-gnu`, then the Intel compiler is able to find files in `/usr/include/x86_64-linux-gnu`. I cannot speculate on how `icc` finds their include directories, but it seems something like:
```
triplet=$(gcc -### | grep Target | awk '{print $2}')
if [ -f /usr/include/${triplet} ] ; then 
    CPATH+=:/usr/include/${triplet}
fi
```

So, this helps to get Intel compiler installed in those two cases and makes `gcc` behave more like a system installed one.

Commit:
Addressing @michaelkuhn suggestions

We can explicitly ask for gcc as compiler. This makes it failsafe in the case
that someone builds gcc with another compiler.

Also, adding `--host` for completeness.

stephenmsachs commented:
> Should we wrap this in `if spec.satisfies("%gcc")` and use `self.compiler.cc` to be extra sure (since `gcc` might not actually be the compiler we build GCC with)?
> 
> Moreover, shouldn't we also set `--host` to the same value (see https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html)?

Yes, that sounds reasonable. The ubuntu system installed `gcc` will actually have all three defined while the RHEL one is missing `--host`. I don't think there is any harm in adding it. I tested it to still work as expected wih the Intel compiler.

haampie commented:
FWIW, the source of this value is `bash -c 'printf "$MACHTYPE"'`. I think bash might also be a missing dependency of GCC (also requires fixing shebangs)

stephenmsachs commented:
@alalazo I removed the implicit assumption that `gcc` is compiled with `gcc` and added `if spec.satisfies("%gcc"):`. Are there any more changes necessary?

alalazo commented:
@stephenmsachs Can you confirm this issue is only with `icc` and not with `icx`? I tried to reproduce on my Ubuntu laptop, and `icx` seems to work fine. I'm asking since `icc` is being [deprecated very soon](https://community.intel.com/t5/Intel-oneAPI-IoT-Toolkit/DEPRECATION-NOTICE-Intel-C-Compiler-Classic/m-p/1412267)

alalazo commented:
![Screenshot from 2023-10-11 21-22-43](https://github.com/spack/spack/assets/4199709/1f00eece-d56b-485e-926f-79988df9dfc3)

stephenmsachs commented:
> @stephenmsachs Can you confirm this issue is only with `icc` and not with `icx`? I tried to reproduce on my Ubuntu laptop, and `icx` seems to work fine. I'm asking since `icc` is being [deprecated very soon](https://community.intel.com/t5/Intel-oneAPI-IoT-Toolkit/DEPRECATION-NOTICE-Intel-C-Compiler-Classic/m-p/1412267)

@alalazo I can confirm the issue is only  with `icc` not `icx`. However, many users are still going to be using `icc` going forward for some time.

Also, this PR makes the spack installed `gcc` act more like the system installed one, which I assume is the 'correct' way as intended by the OS developers. As I said [here](https://github.com/spack/spack/pull/40402#issue-1934863725), for Debian based systems the output of `sh config.guess` is the 'wrong' target for gcc.

alalazo commented:
The concern I have with this PR is that it seems more of a hack than a complete solution. 

> Also, this PR makes the spack installed gcc act more like the system installed one, which I assume is the 'correct' way as intended by the OS developers.

This is not entirely true, since this PR makes the built gcc having the same build, host, target as the `gcc` used to compile it. Nothing ensures that it is the system compiler (and using a manually built gcc wouldn't be so unusual on some clusters).

Even then, why do you say `config.guess` is 'wrong', and you rule out the possibility that it's `icc` to be wrong in how it deals with `gcc`? I may have missed a reference there.

stephenmsachs commented:
> The concern I have with this PR is that it seems more of a hack than a complete solution.

I believe it's a complete solution if we look separately for `--build` and `--host`flags on the `%gcc` and use those separately if theyr differ from the `--target` argument.

> This is not entirely true, since this PR makes the built gcc having the same build, host, target as the `gcc` used to compile it. Nothing ensures that it is the system compiler (and using a manually built gcc wouldn't be so unusual on some clusters).

I see your point. But would that not be even more of a reason to use the same host/target/build? If someone changed those configure arguments on their system they would like to have spack installed `gcc`s behave the same way.

> Even then, why do you say `config.guess` is 'wrong', and you rule out the possibility that it's `icc` to be wrong in how it deals with `gcc`? I may have missed a reference there.

This is independent of `icc`. In the source tarballs of `gcc` build for Debian/ubuntu the configure args are defined in `debian/rules2` as:
```
CONFARGS += \
	--build=$(DEB_BUILD_GNU_TYPE) \
	--host=$(DEB_HOST_GNU_TYPE) \
	--target=$(TARGET_ALIAS)
```
and 
```
TARGET_ALIAS := $(DEB_TARGET_GNU_TYPE)
```
in `debian/rules.def` (in case we are not cross compiling). 
On a x86_64 ubuntu2204 `dpkg-architecture` reports:
```
DEB_BUILD_GNU_TYPE=x86_64-linux-gnu
DEB_HOST_GNU_TYPE=x86_64-linux-gnu
DEB_TARGET_GNU_TYPE=x86_64-linux-gnu
```

alalazo commented:
I'm trying to get more insight on this, and understand if we have:
1. Other ways to work around the issue
2. Ways to tie the build, host, target to some system default rather than to `%gcc`

Talking about 1, this works too on Debian based systems:
```console
$ icc -v -gcc-name=$(which gcc) -isystem/usr/include/x86_64-linux-gnu hello.c
```
and doesn't require modifications to the GCC recipe. It works no matter the compiler used to build the GCC used by Intel. In `compilers.yaml` I added:
```yaml
- compiler:
    spec: intel@=2021.10.0
    paths:
      cc: /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/intel-oneapi-compilers-2023.2.1-jgluezlclwkbbcfkdilqgua32wpihkxj/compiler/2023.2.1/linux/bin/intel64/icc
      cxx: /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/intel-oneapi-compilers-2023.2.1-jgluezlclwkbbcfkdilqgua32wpihkxj/compiler/2023.2.1/linux/bin/intel64/icpc
      f77: /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/intel-oneapi-compilers-2023.2.1-jgluezlclwkbbcfkdilqgua32wpihkxj/compiler/2023.2.1/linux/bin/intel64/ifort
      fc: /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/intel-oneapi-compilers-2023.2.1-jgluezlclwkbbcfkdilqgua32wpihkxj/compiler/2023.2.1/linux/bin/intel64/ifort
    flags:
      cflags: "-gcc-name=/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/gcc-13.2.0-pqx4pqw4kabnmqeb2c77tw4ow5w4qfbj/bin/gcc -isystem/usr/include/x86_64-linux-gnu"
    operating_system: ubuntu20.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```
and I could perform some simple compilations (e.g. `zlib`).

alalazo commented:
Another point is that this:
```console
$ gcc -print-multiarch
```
seems more consistent than deriving the build, host etc. from `gcc -###` on the `Target:` line.

stephenmsachs commented:
I think you are correct with that `-print-multiarch` argument. `gcc` on ubuntu will always report `x86_64-linux-gnu` independent of the build target, and `gcc` on RHEL derivates report an empty string.

I opened https://github.com/spack/spack/pull/40481 to fix only `icc` and leave `gcc` as is.

stephenmsachs commented:
Root cause for `icc` is solved in https://github.com/spack/spack/pull/40481
